common/djangoapps/track/transformers.py

```diff
@@ -0,0 +1,477 @@
+
+import json
+import logging
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import UsageKey
+
+log = logging.getLogger(__name__)
+
+
+class DottedPathMapping(object):
+    """
+    Dictionary-like object for creating keys of dotted paths.
+
+    If a key is created that ends with a dot, it will be treated as a path
+    prefix.  Any value whose prefix matches the dotted path can be used
+    as a key for that value, but only the most specific match will
+    be used.
+    """
+
+    def __init__(self, registry=None):
+        self._match_registry = {}
+        self._prefix_registry = {}
+        self.update(registry or {})
+
+    def __contains__(self, key):
+        try:
+            _ = self[key]
+            return True
+        except KeyError:
+            return False
+
+    def __getitem__(self, key):
+        if key in self._match_registry:
+            return self._match_registry[key]
+        if isinstance(key, basestring):
+            for prefix in sorted(self._prefix_registry, reverse=True):
+                if key.startswith(prefix):
+                    return self._prefix_registry[prefix]
+        raise KeyError('Key {} not found in {}'.format(key, type(self)))
+
+    def __setitem__(self, key, value):
+        if key.endswith('.'):
+            self._prefix_registry[key] = value
+        else:
+            self._match_registry[key] = value
+
+    def __delitem__(self, key):
+        if key.endswith('.'):
+            del self._prefix_registry[key]
+        else:
+            del self._match_registry[key]
+
+    def get(self, key, default=None):
+        """
+        Return `self[key]` if it exists, otherwise, return `None` or `default`
+        if it is specified.
+        """
+        try:
+            self[key]
+        except KeyError:
+            return default
+
+    def update(self, dict_):
+        """
+        Update the mapping with the values in the supplied `dict`.
+        """
+        for key, value in dict_:
+            self[key] = value
+
+    def keys(self):
+        """
+        Return the keys of the mapping, including both exact matches and
+        prefix matches.
+        """
+        return self._match_registry.keys() + self._prefix_registry.keys()
+
+
+class EventTransformerRegistry(object):
+    """
+    Registry to track which EventTransformers handle which events.  The EventTransformer
+    must define a `transformer_name` attribute which contains the name or prefix of the
+    event names it tracks.  Any transformer_name  that ends with a `.` will match all events
+    that share its prefix.  A transformer name without a trailing dot only processes exact
+    matches.
+    """
+
+    mapping = DottedPathMapping()
+
+    @classmethod
+    def register(cls, transformer):
+        """
+        Decorator to register an EventTransformer.  It must have a `transformer_name` class attribute
+        defined.
+        """
+        cls.mapping[transformer.transformer_name] = transformer
```

Comment:
2 comments:
* `transformer_name` can sufficely just be `name` since `transformer` is implied as the container of the field.
* Is it really the "name"?  Rather, should it be something like "match_key"?